062-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,202 @@
+
+PoC: https://github.com/mimaison/prometheus-metrics-reporter
+
+
+## Current situation
+
+Metrics are a critical aspect of monitoring Kafka. Today, this is the way metrics from Kafka are collected. First Kafka creates metrics objects in memory. For historical reasons Kafka uses two different libraries for metrics:
+- a home grown library, `org.apache.kafka.common.metrics`. We’ll refer to metrics created via this as _KafkaMetrics_. This is used on the client side and for common metrics on the server side too.
+- Yammer, `com.yammer.metrics.metrics-core`. This library is the predecessor of [Dropwizard](https://metrics.dropwizard.io/). We’ll refer to metrics created via this as _YammerMetrics_. This is only used on the broker side.
+
+For each type of metrics, Kafka exposes a reporter interface to expose metrics to monitoring systems. KafkaMetrics use [org.apache.kafka.common.metrics.MetricsReporter](https://kafka.apache.org/36/javadoc/org/apache/kafka/common/metrics/MetricsReporter.html) and YammerMetrics use `kafka.metrics.KafkaMetricsReporter` which is not officially part of the public API. Kafka has built-in metrics reporter implementations for JMX for both types.
+
+At the moment Strimzi relies on these default JMX reporters and uses [jmx_exporter](https://github.com/prometheus/jmx_exporter) which is a Java agent that retrieves metrics via JMX and exposes them over an HTTP endpoint in the prometheus format. Then Prometheus is configured to scrape that endpoint to retrieve the Kafka metrics.
+
+![Current Situation](images/062-current.png)
+
+- `org.apache.kafka.common.metrics.JmxReporter` is the reporter implementation for the Kafka Metrics.
+- `org.apache.kafka.server.metrics.FilteringJmxReporter` is the reporter implementation for the Yammer Metrics. It’s named `FilteringJmxReporter` because it extends the JmxReporter class from the Yammer library and adds an option to select the metrics to report.
+
+
+## Motivation
+
+I propose updating the metrics collection pipeline for the following reasons:
+
+1. The current metrics collection pipeline is pretty convoluted. We have metrics reporters first exposing metrics via JMX before using a Java agent to expose them again via HTTP. Using metrics reporters to directly expose metrics to Prometheus would significantly simplify it by removing JMX from the picture and removing jmx_exporter.
+2. Each component along the pipeline has its own configurations and specificities. For example, since Kafka 3.4.0, it’s possible to disable the JmxReporter for KafkaMetrics by setting [auto.include.jmx.reporter](https://kafka.apache.org/documentation/#brokerconfigs_auto.include.jmx.reporter) to `false`. It is currently not possible to disable `FilteringJmxReporter` (it’s something I plan to address in the near future). Also both jmx_exporter and the built-in reporter allow selecting metrics to collect.
```

Comment:
`it’s something I plan to address in the near future` is this still valid? already done? planned?

Reply:
Yes that statement is still valid. However it does not affect the proposal so if it's confusing I can remove it.

Reply:
I think it doesn't add much value to the proposal.

Reply:
Right, I've removed that statement.